### PR TITLE
Fix: Install rules before validate

### PR DESCRIPTION
**Pull request by qzhou1607**
## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [X] Added/Updated related tests.

## Short description of the change(s)
In the current code, if a rule has not been installed, when user scans a site, an error is thrown and the user will be asked to re-generate a valid `.sonarwhalrc` even though there already is one. The correct behavior should be prompt to ask the user if they want to install the missing rules first. So this PR moves the validation of rule configurations after the rules have been installed.

### CODE REVIEW

File: packages/sonarwhal/src/lib/config.ts [hunk 1]
```diff
@@ -190,16 +190,19 @@ const loadIgnoredUrls = (userConfig: UserConfig): Map<string, RegExp[]> => {
 /** Validates that the given configuration for a rule is valid */
 const validateRules = (rulesConfig: RulesConfigObject, userConfig: UserConfig) => {
     const rules = Object.keys(rulesConfig);
+    const invalidRuleConfigs = [];
```
Comment (molant) [inline packages/sonarwhal/src/lib/config.ts]: Can we have a `reduce` here?

Comment (qzhou1607) [reply packages/sonarwhal/src/lib/config.ts]: @molant Done.

File: packages/sonarwhal/src/lib/config.ts [hunk 2]
```diff
@@ -190,16 +190,18 @@ const loadIgnoredUrls = (userConfig: UserConfig): Map<string, RegExp[]> => {
...
         if (!valid) {
-            throw new Error(`Rule ${rule} has an invalid configuration`);
+            invalidRules.push(rule);
         }
-    });
+
+        return invalidRules;
+    }, []);
+
+    return invalidRuleConfigs;
```
Comment (molant) [inline packages/sonarwhal/src/lib/config.ts]: I think I would change the return of this to be more aligned with what we do in other places and also so the name of the function is still valid. Maybe return something like:
```js
{
  valid: [],
  invalid: []
}
```

File: packages/sonarwhal/src/lib/config.ts [hunk 3]
```diff
@@ -265,11 +267,19 @@ export class SonarwhalConfig {
...
-        validateRules(rules, userConfig);
-
         return new SonarwhalConfig(userConfig, browsers, ignoredUrls, normalizeRules(rules));
     }
 
+    /**
+     * Return a list of rules that have invalid configurations.
+     * @param config
+     */
+    public static validateRuleConfig(config: UserConfig): Array<string> {
```
Comment (molant) [inline packages/sonarwhal/src/lib/config.ts]: Shouldn't this be `validateRulesConfig`?

Comment (molant) [reply packages/sonarwhal/src/lib/config.ts]: Also, do we need to normalize here? I thought this was done earlier in the process.

**Pull request merged**